src/c_api.cpp

```diff
@@ -1594,6 +1597,59 @@ int LGBM_BoosterLoadModelFromString(
   API_END();
 }
 
+// extracts the value for key in a parameter line. E.g. for the key boosting and
+// a line like `[boosting: gbdt]` (without newline character), the extracted value would be "gbdt".
+std::string extract_param(std::string key, std::string line) {
+  std::string prefix = "[" + key + ": ";
+  std::string suffix = "]";
+  std::string prefix_replaced = line.replace(line.begin(), line.begin() + prefix.length(), "");
+  std::string suffix_replaced = prefix_replaced.replace(prefix_replaced.end() - suffix.length(), prefix_replaced.end(), "");
+  return suffix_replaced;
+}
+
+int LGBM_BoosterGetConfig(
+  BoosterHandle handle,
+  int64_t buffer_len,
+  int64_t* out_len,
+  char* out_str
+) {
+  API_BEGIN();
+  Booster* ref_booster = reinterpret_cast<Booster*>(handle);
```

Comment:
I'd like to propose a different organization of this code, but please don't make any changes to the organization until another maintainer like @StrikerRUS, @shiyu1994, or @tongwu-msft  comments.

Instead of all this logic being in `src/c_api.cpp`, I think most of the implementation for `LGBM_BoosterGetConfig()` should live somewhere else. Similar to how `LGBM_BoosterSaveModelToString()` is defined here in `src/c_api.cpp`.

https://github.com/microsoft/LightGBM/blob/cf38071b6ac7e238b1654286dbb5a9a98b2c320b/src/c_api.cpp#L2233

And then it is just a thin wrapper on `GBDT::SaveModelToString()`.

https://github.com/microsoft/LightGBM/blob/a06fadfb7ac3fdb26da3d4afc061a8b976070c50/src/boosting/gbdt_model_text.cpp#L410

I also think it's desirable for any code that needs to list every parameter by name should be in the automatically-generated methods in `src/config_auto.cpp`.

So I think this should be approached as following:

1. Add a method like `GBDT::GetLoadedConfig()` which reformats `loaded_parameter_` into the appropriate format a parameters string and then initializes a `Config` object by calling `Config::GetMembersFromString()`
    - https://github.com/microsoft/LightGBM/blob/cf38071b6ac7e238b1654286dbb5a9a98b2c320b/src/io/config_auto.cpp#L321
2. Add a method in `src/io/config_auto.cpp` like `Config::DumpConfigToJson()`, which returns a JSON representation of a `Config` object.
    - NOTE: all the code there is generated by https://github.com/microsoft/LightGBM/blob/a06fadfb7ac3fdb26da3d4afc061a8b976070c50/helpers/parameter_generator.py. But if I was working on this, my approach would be to just write code directly in `src/io/config_auto.cpp`, get the tests into a good state, and *then* make `parameter_generator.py` write that code.
    - this will be useful if LightGBM adopts the proposal at #4887 in the future
3. Alter `LGBM_BoosterGetConfig()` to get the output of `Config::DumpConfigToJson()`.

--------------

Sorry if this is overly complicated. Let's see what other maintainers say. And we are here to help! Thanks again for your help with this feature.

Reply:
>reformats `loaded_parameter_` into the appropriate format a parameters string

@jameslamb , do you actually mean reformatting `loaded_parameter_` into a `std::unordered_map<std::string, std::string>` object? 

I'm a bit confused by the name `GetMembersFromString`, it actually means GetMembersFrom an unordered_map, right?

Reply:
I mean doing whatever is necessary to turn `loaded_parameter_` into a `Config` object, yes. You're right that the method `Config::GetMembersFromString()` takes an unordered map with string keys and string values.